### PR TITLE
Inference | Hybrid prefix caching.

### PR DESCRIPTION
**Pull request by lmcafee-nvidia**
# What does this PR do ?

Add Mamba state prefix caching for hybrid Transformer-Mamba models, enabling KV cache prefix sharing to also share corresponding Mamba conv/SSM states.

### Key Features

- **Mamba state caching**: When KV cache blocks are reused due to shared prefixes, the corresponding Mamba states can now also be cached and restored, avoiding redundant recomputation
- **Memory-budgeted cache**: New `--inference-dynamic-batching-prefix-caching-mamba-gb` argument controls the memory budget for cached Mamba states
- **LRU eviction**: Mamba cache uses LRU eviction based on KV block timestamps when the cache is full
- **Automatic invalidation**: When KV blocks are evicted, their associated Mamba states are automatically invalidated

### Changes

**Core Implementation** (`megatron/core/inference/contexts/dynamic_context.py`, `megatron/core/inference/engines/dynamic_engine.py`):
- Fixed-size tensor pool for cached Mamba conv/SSM states
- Block-to-slot mapping for associating KV blocks with Mamba cache slots
- Store Mamba state at block boundaries after prefill completes
- Restore cached Mamba state when prefix matches during request scheduling

**Block Allocator** (`megatron/core/inference/contexts/dynamic_block_allocator.py`):
- Invalidate Mamba state when KV blocks are evicted

**Arguments** (`megatron/training/arguments.py`):
- Add `--inference-dynamic-batching-prefix-caching-mamba-gb` parameter
- Auto-enable chunked prefill when Mamba prefix caching is used

## Test plan

- [x] Unit tests for cache allocation/deallocation
- [x] Unit tests for LRU eviction (20 tests in `tests/unit_tests/inference/contexts/test_mamba_prefix_caching.py`)
- [x] Tests for memory budget edge cases
- [x] Tests for state store/restore integration
- [ ] End-to-end inference test with hybrid model

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

Feel free to message or comment the @mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/ssm/mamba_mixer.py [hunk 1]
```diff
@@ -504,68 +527,120 @@ def _dynamic_inference_prefill(
...
+        # Strip CUDA-graph padding from metadata tensors. Padded
+        # entries have -1 batch_indices and zero-length cu_seqlens segments,
+        # which cause out-of-bounds indexing in the varlen SSM kernel.
+        # Also strip padded tokens from zxBCdt to keep all downstream tensor
+        # shapes (z, xBC, dt) consistent. Pad output back afterward for
+        # residual add compatibility.
+        cu_seqlens = metadata.cu_seqlens[: real_prefill_count + 1]
+        batch_indices = metadata.batch_indices_prefill[:real_prefill_count]
+        real_token_count = cu_seqlens[-1].item()
+        seq_idx = metadata.seq_idx[:, :real_token_count] if metadata.seq_idx is not None else None
```
Comment (santhnm2) [inline megatron/core/ssm/mamba_mixer.py]: Why do we need to do this?

File: megatron/core/ssm/mamba_mixer.py [hunk 2]
```diff
@@ -504,68 +527,120 @@ def _dynamic_inference_prefill(
...
+                            torch.full((slen,), offset, dtype=seq_idx.dtype, device=seq_idx.device)
+                        )
+                        offset += 1
+                seq_idx = torch.cat(seq_idx_vals).unsqueeze(0) if seq_idx_vals else None
+            if intermediate_token_offsets is not None:
+                intermediate_token_offsets = [
+                    intermediate_token_offsets[i]
+                    for i in range(real_prefill_count)
+                    if nonzero_mask[i]
+                ]
```
Comment (santhnm2) [inline megatron/core/ssm/mamba_mixer.py]: Can this be moved to `mamba_metadata.update`?

File: megatron/core/ssm/mamba_mixer.py [hunk 3]
```diff
@@ -504,68 +527,120 @@ def _dynamic_inference_prefill(
...
+                intermediate_states = expanded
         else:
-            return y_regular
+            y_prefill = result
+            intermediate_states = None
+
+        # Pad output back to padded token count. The caller (residual add,
+        # tensor_merge) expects the output to match the padded input shape.
+        if y_prefill.shape[0] < padded_token_count:
+            y_prefill = F.pad(y_prefill, (0, 0, 0, 0, 0, padded_token_count - y_prefill.shape[0]))
```
Comment (santhnm2) [inline megatron/core/ssm/mamba_mixer.py]: Can we do `torch.empty` and `tensor_merge` instead to be cuda-graphable?

File: megatron/core/ssm/mamba_mixer.py [hunk 4]
```diff
@@ -708,31 +782,55 @@ def _ssm_prefill(
...
+                init_r = init_r.permute(0, 2, 1).contiguous().transpose(1, 2)  # channels-last
+                xBC_r = causal_conv1d_fn(
+                    x=xBC_r,
+                    weight=conv_weight,
+                    bias=conv_bias,
+                    activation=self.activation,
+                    initial_states=init_r,
+                )
+                xBC_parts.append(xBC_r.transpose(1, 2).contiguous())  # (1, L, C)
+            xBC = torch.cat(xBC_parts, dim=1)  # (1, total_tokens, conv_dim)
```
Comment (santhnm2) [inline megatron/core/ssm/mamba_mixer.py]: This won't be cuda graphable right? Is it possible to move this loop into the kernel?

File: megatron/core/ssm/mamba_mixer.py [hunk 5]
```diff
@@ -774,66 +863,167 @@ def _ssm_prefill(
...
         x = rearrange(x, "b l (h p) -> b l h p", p=self.headdim).contiguous()
         dt = dt.contiguous()
         B = rearrange(B, "b l (g n) -> b l g n", n=self.d_state).contiguous()
         C = rearrange(C, "b l (g n) -> b l g n", n=self.d_state).contiguous()
         z = rearrange(z, "b l (h p) -> b l h p", p=self.headdim).contiguous()
 
-        # If `rmsnorm == False`, then the norm inside `mamba_chunk_scan_combined` will be used.
-        # In this case, if `cp_size > 1` then that norm could be performed on less heads than if
-        # `cp_size == 1` (groups of heads can be sharded across CP ranks), which would be
-        # mathematically incorrect, and potentially arithmetically unstable.
```
Comment (santhnm2) [inline megatron/core/ssm/mamba_mixer.py:811]: Why did we remove this comment?

File: megatron/core/ssm/mamba_mixer.py [hunk 6]
```diff
@@ -774,66 +863,167 @@ def _ssm_prefill(
...
+            y = torch.empty_like(x)
+
+            # Build cu_chunk_seqlens by subdividing each sequence into
+            # chunks of at most self.chunk_size tokens. The SSM Triton kernels
+            # allocate per-chunk output arrays of size chunk_size, so passing
+            # cu_seqlens directly would cause out-of-bounds memory access when
+            # any sequence exceeds chunk_size (default 128) tokens.
+            # Also build last_chunk_indices: the index of the last chunk for
+            # each sequence, used to extract final SSM states directly from
+            # the states tensor without a separate kernel call.
```
Comment (santhnm2) [inline megatron/core/ssm/mamba_mixer.py]: Can this code block also be moved into `mamba_metadata.update`?

File: megatron/core/inference/contexts/dynamic_context.py [hunk 7]
```diff
@@ -959,6 +978,374 @@ def mamba_states_cache(self, layer_number: int) -> Tuple[Tensor, Tensor]:
...
+        num_blocks = self.block_allocator.total_count
+        self.block_to_mamba_slot = torch.full((num_blocks,), -1, dtype=torch.int32, device=device)
+        self.mamba_slot_to_block = torch.full((max_slots,), -1, dtype=torch.int32, device=device)
+
+        # Free slot pool (stack)
+        self.mamba_cache_free_slots = torch.arange(max_slots, dtype=torch.int32, device=device)
+        self.mamba_cache_free_count = max_slots
+
+        # State tensors
+        self.mamba_cache_conv_states = torch.zeros(
```
Comment (santhnm2) [inline megatron/core/inference/contexts/dynamic_context.py]: Just to differentiate from the MTP intermediate states that will also be added, can we explicitly call these `mamba_prefix_cache_free_slots`?

File: megatron/core/inference/contexts/dynamic_context.py [hunk 8]
```diff
@@ -959,6 +978,374 @@ def mamba_states_cache(self, layer_number: int) -> Tuple[Tensor, Tensor]:
...
+        device = torch.cuda.current_device()
+
+        # Block <-> slot mappings
+        num_blocks = self.block_allocator.total_count
+        self.block_to_mamba_slot = torch.full((num_blocks,), -1, dtype=torch.int32, device=device)
+        self.mamba_slot_to_block = torch.full((max_slots,), -1, dtype=torch.int32, device=device)
+
+        # Free slot pool (stack)
+        self.mamba_cache_free_slots = torch.arange(max_slots, dtype=torch.int32, device=device)
+        self.mamba_cache_free_count = max_slots
```
Comment (santhnm2) [inline megatron/core/inference/contexts/dynamic_context.py]: Should these go in `MambaMetadata?

File: megatron/core/ssm/mamba_mixer.py (hunk 1 above)
Comment (lmcafee-nvidia) [reply megatron/core/ssm/mamba_mixer.py]: this stripping was removed while working on cuda graph compatibility. the precomputed metadata in `MambaMetadata.update()` handles the padding

File: megatron/core/ssm/mamba_mixer.py (hunk 2 above)
Comment (lmcafee-nvidia) [reply megatron/core/ssm/mamba_mixer.py]: the zero-length filtering code has been completely removed. we recently switched from a system that resulted in zero-length prefills, to a system the recomputes at least 1+ tokens, so this code is no longer around

File: megatron/core/ssm/mamba_mixer.py (hunk 3 above)
Comment (lmcafee-nvidia) [reply megatron/core/ssm/mamba_mixer.py]: this has been updated to be cuda-graphable, I've tested the entire forward pass when using the new Triton conv1d kernel. F.pad was removed, and now the full paded tensor flows end-to-end

File: megatron/core/ssm/mamba_mixer.py (hunk 4 above)
Comment (lmcafee-nvidia) [reply megatron/core/ssm/mamba_mixer.py]: so we have two paths through the system, 1) per-request sequential conv loop, and 2) new Triton kernel. The conv-loop is inherently non-cuda-graphable. Before prefix caching, we always started from token 0 so we could batch everything within the varlen kernel. But now with prefix caching, we need to sequentially iterate the batch kernel per request, so we can't cuda graph that.

the Triton kernel is the cuda graph path. when `use_triton_con1d==True`, we batch all convs through there. So, we do have at least one end-to-end cuda-graphable path.

File: megatron/core/ssm/mamba_mixer.py (hunk 6 above)
Comment (lmcafee-nvidia) [reply megatron/core/ssm/mamba_mixer.py]: this was indeed moved into `MambaMetadata.update()` so that everything is precomputed

File: megatron/core/inference/contexts/dynamic_context.py (hunk 7 above)
Comment (lmcafee-nvidia) [reply megatron/core/inference/contexts/dynamic_context.py]: all of this slot management code was moved into a new `MambaSlotAllocator` class (`./mamba_slot_allocator.py`), so hopefully there won't be any name clash

File: megatron/core/inference/contexts/dynamic_context.py (hunk 8 above)
Comment (lmcafee-nvidia) [reply megatron/core/inference/contexts/dynamic_context.py]: these were moved into the new `MambaSlotAllocator` class. the difference between this and the metadata class is that the metadata is generally focused on per-request data, and the slot allocator is generally focused on prefix caching slots. there's a bit of overlap in some variables, but that's the general division of responsibility.

File: megatron/core/ssm/mamba_mixer.py (hunk 5 above)
Comment (lmcafee-nvidia) [reply megatron/core/ssm/mamba_mixer.py:811]: restored both comment groups, they were mistakingly removed

File: megatron/core/ssm/mamba_mixer.py [hunk 9]
```diff
@@ -131,7 +144,7 @@ class MambaMixer(MegatronModule):
         dt_init_floor: The minimum value of the dt parameter after initialization.
         bias: Whether to use bias in the linear layers.
         conv_bias: Whether to use bias in the causal convolution.
-        chunk_size: The chunk size for the fused kernel.
+        mamba_chunk_size: The chunk size for the Mamba SSM fused kernel.
```
Comment (santhnm2) [inline megatron/core/ssm/mamba_mixer.py]: Undo this change since we just need to set `mamba_chunk_size` in `megatron/core/inference/config.py`

**Review by kvareddy — APPROVED** [review]

**Review by shanmugamr1992 — APPROVED** [review]

File: megatron/core/inference/engines/dynamic_engine.py [hunk 10]
```diff
@@ -603,7 +633,9 @@ async def start_listening_to_data_parallel_coordinator(
         # initialize zmq-based world communicator for consensus barriers
         total_world_size = torch.distributed.get_world_size()
         if total_world_size > 1:
-            self.world_zmq_communicator = AsyncZMQCommunicator(self.zmq_context, process_group=None)
+            self.world_zmq_communicator = AsyncZMQCommunicator(
+                self.zmq_context, process_group=torch.distributed.group.WORLD
```
Comment (santhnm2) [inline megatron/core/inference/engines/dynamic_engine.py]: Try without this change and see if this works?

File: megatron/core/inference/engines/dynamic_engine.py [hunk 11]
```diff
@@ -1656,6 +1737,15 @@ async def async_bookkeep(
             self.socket_for_receiving_requests.send(payload)
             range_pop()
 
+        # Drain prefix cache hit counters from context into engine accumulators.
+        if self.context.enable_prefix_caching:
+            self._prefix_cache_hits += self.context.prefix_cache_hits
+            self._prefix_cache_blocks_matched += self.context.prefix_cache_blocks_matched
+            self._prefix_cache_tokens_skipped += self.context.prefix_cache_tokens_skipped
```
Comment (santhnm2) [inline megatron/core/inference/engines/dynamic_engine.py]: Make sure this is removed (same with line below)

File: megatron/core/inference/engines/dynamic_engine.py [hunk 12]
```diff
@@ -1685,6 +1775,16 @@ async def async_bookkeep(
                 metrics['inference/spec_decode_tokens_accepted'] = int(self._spec_tokens_accepted)
                 metrics['inference/spec_decode_num_steps'] = int(self._spec_steps)
 
+            # Add prefix caching metrics.
+            if self.context.enable_prefix_caching and self._prefix_cache_hits > 0:
+                metrics['inference/prefix_cache_hits'] = int(self._prefix_cache_hits)
+                metrics['inference/prefix_cache_blocks_matched'] = int(
+                    self._prefix_cache_blocks_matched
+                )
+                metrics['inference/prefix_cache_tokens_skipped'] = int(
```
Comment (santhnm2) [inline megatron/core/inference/engines/dynamic_engine.py]: Make sure this is removed

**Review by santhnm2 — APPROVED** [review]

**Review by asolergi-nv — APPROVED** [review]

**Review by jaredcasper — APPROVED** [review]

**Review by jiemingz — APPROVED** [review]

**Review by chtruong814 — APPROVED** [review]
